cpp/src/arrow/csv/parser.cc

```diff
@@ -171,12 +186,26 @@ class ResizableValueDescWriter : public ValueDescWriter<ResizableValueDescWriter
 // faster CSV parsing code.
 class PresizedValueDescWriter : public ValueDescWriter<PresizedValueDescWriter> {
  public:
+  // The number of offsets being written will be `1 + num_rows * num_cols`,
+  // however we allow for one extraneous write in case of excessive columns,
+  // hence `2 + num_rows * num_cols` (see explanation in PushValue below).
   PresizedValueDescWriter(MemoryPool* pool, int32_t num_rows, int32_t num_cols)
-      : ValueDescWriter(pool, /*values_capacity=*/1 + num_rows * num_cols) {}
+      : ValueDescWriter(pool, /*values_capacity=*/2 + num_rows * num_cols) {}
 
   void PushValue(ParsedValueDesc v) {
     DCHECK_LT(values_size_, values_capacity_);
-    values_[values_size_++] = v;
+    values_[values_size_] = v;
+    // We must take care not to write past the buffer's end if the line being
+    // parsed has more than `num_cols` columns. The obvious solution of setting
```

Comment:
Could you elaborate a bit on the "obvious solution of setting an error status"?

Reply:
Well, the intuitive approach is to return a `Status` from this function, or to let it accumulate in `status_`. But it turns out that hurts performance too much (the CSV parser is a state machine with a huge transition rate - realistically once every few bytes of input).

Reply:
(as a side note, this is probably a good argument for C++ exceptions generally, since they don't incur any cost in the happy case - unlike explicit error returns)

Reply:
> (as a side note, this is probably a good argument for C++ exceptions generally, since they don't incur any cost in the happy case - unlike explicit error returns)

And now that remark gives me an idea for solving this without any performance cost. It's gonna be hackish though :(

Reply:
Something like below?
```
// The error path.
if unlikely(values_size_ >= values_capacity_) {
  status_ &= Status::Invalid("xxx");
  return;
}
// The valid path.
...
```

And this hurts perf if so?

Reply:
I tested something like that already and, yes, it hurts by ~30% on parser benchmarks IIRC.
(current solution hurts by ~10-15%)

Reply:
Hmm, that's surprising, how can a unlikely error branching hurt perf so bad...

Reply:
(might be all three of course!)

Reply:
Does the error branch never happen in the benchmark (and still perf is down), or is the benchmark testing bad cases as well?

Reply:
The benchmark is only testing good cases (we don't care about the performance of error reporting).

Reply:
Of course it might depend on the machine and compiler...

Reply:
Got it, that's really something interesting to know. Thank you.